### PR TITLE
feat: add custom token logic

### PR DESCRIPTION
**Pull request by berteotti**
- Adds custom token logic to Token Picker. 
- Made some changes to Token Picker logic.

image

### CODE REVIEW

File: packages/app/components/token-picker/TokenPicker.tsx [hunk 1]
```diff
@@ -46,21 +53,38 @@ export const TokenPicker = ({
...
 
+  const tokenListCleanup = () => {
+    setFilteredTokenList(
+      isConnected && tokenListWithBalances ? tokenListWithBalances : tokenList
+    );
+  };
+
   const handleTokenSearchInput = (event: ChangeEvent<HTMLInputElement>) => {
-    setDebouncedQuery(event.target.value);
+    const query = event.target.value;
```
Comment (ElRodrigote) [inline packages/app/components/token-picker/TokenPicker.tsx:82]: Maybe we can trim this value to take out the empty spaces from the query, WDYT?

image

File: packages/app/components/token-picker/TokenPicker.tsx [hunk 2]
```diff
@@ -46,21 +53,38 @@ export const TokenPicker = ({
...
+    setDebouncedQuery(query);
+
+    if (query === "") {
```
Comment (ElRodrigote) [inline packages/app/components/token-picker/TokenPicker.tsx:85]: **MINOR, non-blocking:**

```suggestion
    if (!query) {
```

File: packages/app/components/token-picker/TokenPicker.tsx [hunk 3]
```diff
@@ -93,6 +117,70 @@ export const TokenPicker = ({
...
+          token.name.toLowerCase().includes(debouncedQuery.toLowerCase())
+      );
+
+      setFilteredTokenList(filteredList);
+    }
+  }, [isConnected, tokenList, tokenListWithBalances, debouncedQuery]);
+
+  useEffect(() => {
+    (async () => {
+      if (isAddress(debouncedQuery) && filteredTokenList.length === 0) {
```
Comment (ElRodrigote) [inline packages/app/components/token-picker/TokenPicker.tsx:152]: **MINOR, non-blocking:**

```suggestion
      if (isAddress(debouncedQuery) && !filteredTokenList.length) {
```

File: packages/app/components/token-picker/TokenPicker.tsx [hunk 4]
```diff
@@ -171,85 +260,140 @@ interface TokenListProps {
...
+          <DialogUnknownToken
+            isOpen={isDialogOpen}
+            closeAction={() => setIsDialogOpen(false)}
+            confirmAction={() => {
+              setIsDialogOpen(false);
+              onTokenSelect(token);
+            }}
+            token={token}
+          />
+        )}
```
Comment (ElRodrigote) [inline packages/app/components/token-picker/TokenPicker.tsx:354]: **MINOR, non-blocking:**

```suggestion
        {token.isImported && (
          <>
            <BodyText>Import token</BodyText>
            <DialogUnknownToken
              isOpen={isDialogOpen}
              closeAction={() => setIsDialogOpen(false)}
              confirmAction={() => {
                setIsDialogOpen(false);
                onTokenSelect(token);
              }}
              token={token}
            />
          </>
        )}
```

File: packages/app/components/token-picker/TokenPicker.tsx [hunk 5]
```diff
@@ -93,6 +117,70 @@ export const TokenPicker = ({
...
+      const newList =
+        isConnected && tokenListWithBalances
+          ? tokenListWithBalances
+          : tokenList;
+
+      const filteredList = newList.filter(
+        (token) =>
+          token.symbol.toLowerCase().includes(debouncedQuery.toLowerCase()) ||
+          token.name.toLowerCase().includes(debouncedQuery.toLowerCase())
+      );
```
Comment (Diogomartf) [inline packages/app/components/token-picker/TokenPicker.tsx:144]: makes sense to filter per address, sometimes people paste the address, it's handy so you are sure it is the token you're looking for.

File: packages/app/components/token-picker/TokenPicker.tsx (hunk 1 above)
Comment (berteotti) [reply packages/app/components/token-picker/TokenPicker.tsx:85]: This is a good UX. Will fix it!

File: packages/app/components/token-picker/TokenPicker.tsx (hunk 2 above)
Comment (berteotti) [reply packages/app/components/token-picker/TokenPicker.tsx]: Done

File: packages/app/components/token-picker/TokenPicker.tsx (hunk 4 above)
Comment (berteotti) [reply packages/app/components/token-picker/TokenPicker.tsx]: Done

File: packages/app/components/token-picker/TokenPicker.tsx (hunk 5 above)
Comment (berteotti) [reply packages/app/components/token-picker/TokenPicker.tsx]: You are right, this was a bug. Fixed

File: packages/app/components/token-picker/TokenPicker.tsx [hunk 6]
```diff
@@ -171,85 +260,140 @@ interface TokenListProps {
...
-            ? token.balance
-            : formatTokenValue(token.balance as string)}
-        </BodyText>
-      )}
     </div>
-  </div>
+  );
+};
+
+interface DialogUnknownTokenProps {
```
[the remark this answers is not in the document]
Comment (berteotti) [reply packages/app/components/token-picker/TokenPicker.tsx]: Done, thanks!

**Comment by Diogomartf** [comment]
image
I would be nice that "Import token" was a button

File: packages/app/components/token-picker/TokenPicker.tsx [hunk 7]
```diff
@@ -171,54 +263,46 @@ interface TokenListProps {
...
+            token={token}
+          />
+        ))
+      ) : (
+        tokenSearchQuery && (
+          <div className="flex flex-col items-center justify-center mt-8 space-y-4">
+            <EmptyStateTokenPickerImg />
+            <BodyText className="text-center">{`Nothing found for "${tokenSearchQuery}".`}</BodyText>
+            {queryIsAddress && !isConnected && (
+              <BodyText className="text-center">
```
Comment (Diogomartf) [inline packages/app/components/token-picker/TokenPicker.tsx:296]:
```suggestion
              <BodyText className=" text-center text-em-med text-sm" 
>
```

Comment (berteotti) [reply packages/app/components/token-picker/TokenPicker.tsx]: Done

File: packages/app/components/token-picker/TokenPicker.tsx [hunk 8]
```diff
@@ -46,21 +52,41 @@ export const TokenPicker = ({
...
+      isConnected && tokenListWithBalances ? tokenListWithBalances : tokenList
+    );
+  };
+
   const handleTokenSearchInput = (event: ChangeEvent<HTMLInputElement>) => {
-    setDebouncedQuery(event.target.value);
+    const query = event.target.value;
+    setDebouncedQuery(query);
+
+    if (!query.trim()) {
```
Comment (Diogomartf) [inline packages/app/components/token-picker/TokenPicker.tsx:87]: why trim here?

Can we just check if there is a query?

**Review by Diogomartf — APPROVED** [review]

File: packages/app/components/token-picker/TokenPicker.tsx [hunk 9]
```diff
@@ -227,29 +311,93 @@ interface TokenListRowProps {
...
+            <Button variant="secondary">Import token</Button>
+            <DialogImportToken
+              isOpen={isDialogOpen}
+              closeAction={() => setIsDialogOpen(false)}
+              confirmAction={() => {
+                setIsDialogOpen(false);
+                onTokenSelect(token);
+              }}
+              token={token}
+            />
```
[the remark this answers is not in the document]
Comment (berteotti) [reply packages/app/components/token-picker/TokenPicker.tsx:353]: No, this is to identify the token the user wants to import

**Pull request merged**